[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer

b-koopman opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68390

Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

Differential Revision: [D32446405](https://our.internmc.facebook.com/intern/diff/D32446405)

b-koopman committed:
[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer

Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer"


Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer"


Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32446405).

vkuzo commented on torch/ao/quantization/fake_quantize.py:
```diff
@@ -126,8 +127,12 @@ def __init__(self, observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=
         self.activation_post_process = observer(**observer_kwargs)
         assert torch.iinfo(self.activation_post_process.dtype).min <= quant_min, 'quant_min out of bound'
         assert quant_max <= torch.iinfo(self.activation_post_process.dtype).max, 'quant_max out of bound'
+        if(self.activation_post_process.qscheme == torch.per_channel_affine_float_qparams):
```
would it work to get the dtype from `activation_post_process.zero_point.dtype`?

b-koopman replied:
If `activation_post_process` is an Observer, it will not have `zero_point` as a member, since this is calculated in `_calculate_qparams`
see: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L283-#L325.

To clean this up, I added a a helper `_is_float_qparams`, similar to the other `qscheme` helper functions.
see: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fake_quantize.py#L20

b-koopman committed:
Update on "[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer"


Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

Differential Revision: [D32446405](https://our.internmc.facebook.com/intern/diff/D32446405)

[ghstack-poisoned]

vkuzo reviewed (approved):
lg!

b-koopman committed:
Update on "[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer"


Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

Differential Revision: [D32446405](https://our.internmc.facebook.com/intern/diff/D32446405)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32446405).

b-koopman committed:
Update on "[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer"


Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

Differential Revision: [D32446405](https://our.internmc.facebook.com/intern/diff/D32446405)

[ghstack-poisoned]

b-koopman committed:
Update on "[quant][embedding qat] Set FakeQuant zeropoint dtype matches observer"


Summary:

Observer zero_point's dtype can be float, in the specific case of `torch.per_channel_affine_float_qparams`.
This change sets FakeQuant's zero_point dtype accordingly.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "embedding"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "embedding"`

Reviewers:

Subscribers:

Tasks:
https://github.com/pytorch/pytorch/issues/61866

Tags:

Differential Revision: [D32446405](https://our.internmc.facebook.com/intern/diff/D32446405)

[ghstack-poisoned]

b-koopman commented:
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32446405).
